### PR TITLE
Refactor AttachmentModal to function component

### PR DESCRIPTION
**Pull request by rayane-d**
### Fixed Issues

$ https://github.com/Expensify/App/issues/16114
PROPOSAL: https://github.com/Expensify/App/issues/16114#issuecomment-1574519879

### Tests

1. Open a chat
2. Upload an Attachment via button or via drag and drop
3. Verify that a modal appears displaying a preview of the file
4. Click send and verify that the file is sent successfully in the discussion

- [x] Verify that no errors appear in the JS console

### QA Steps

1. Open a chat
2. Upload an Attachment via button or via drag and drop
3. Verify that a modal appears displaying a preview of the file
4. Click send and verify that the file is sent successfully in the discussion

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/77965000/b3f154b8-4974-49e4-8927-f104ae91faaa

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/77965000/a6a0d82d-9a80-4e56-bed5-337d3aa547b4

</details>

<details>

<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/77965000/e4c499be-66fa-46c4-bfeb-b11b3613ffa0

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/77965000/a71c383a-e9a2-479f-8082-7261f6e19f83

</details>

<details>

<summary>iOS</summary>

https://github.com/Expensify/App/assets/77965000/c8d529fc-1a7f-4eaa-999b-0c79a9a32fd4

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/77965000/a1cce1e8-2553-4d30-8b29-6f66773874b3

</details>

### CODE REVIEW

**Comment by rayane-djouah** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by rayane-djouah** [comment]
@marcochavezf @rushatgabhane please review this PR

**Comment by rayane-djouah** [comment]
@marcochavezf @rushatgabhane

**Comment by rayane-djouah** [comment]
@marcochavezf @rushatgabhane could you approve and run workflows. thanks.

**Comment by rayane-djouah** [comment]
@marcochavezf @rushatgabhane please review this PR

**Comment by rayane-djouah** [comment]
@marcochavezf @rushatgabhane merge conflict resolved

**Comment by rayane-djouah** [comment]
@marcochavezf @rushatgabhane could you review

**Comment by rayane-djouah** [comment]
@marcochavezf prettier error fixed, could you re-run workflows

**Comment by rayane-djouah** [comment]
@marcochavezf @rushatgabhane

**Comment by marcochavezf** [comment]
Hi @rayane-djouah thank you for the bumps. @rushatgabhane (Contributor+ team member) will review the PR first🙂

**Comment by marcochavezf** [comment]
> @marcochavezf Can you please assign this to me as @rushatgabhane has too much on his plate ATM?

Sure!

**Review by allroundexperts — APPROVED** [review]
LGTM!
cc @marcochavezf

**Review by marcochavezf — APPROVED** [review]
Thanks @rayane-djouah! LGTM

File: src/components/AttachmentModal.js [hunk 1]
```diff
@@ -78,164 +78,153 @@ const defaultProps = {
...
      * Keeps the attachment source in sync with the attachment displayed currently in the carousel.
      * @param {{ source: String, isAuthTokenRequired: Boolean, file: { name: string } }} attachment
      */
-    onNavigate(attachment) {
-        this.setState(attachment);
-    }
+    const onNavigate = useCallback((attachment) => {
+        setSource(attachment.source);
+        setFile(attachment.file);
+    }, []);
```
[the remark this answers is not in the document]
Comment (kidroca) [reply src/components/AttachmentModal.js:107]: Thank you for your input, @allroundexperts. I appreciate your vigilance and the tests you performed before merging the PR. However, I'd like to highlight the way `fileDownload` operates. It relies heavily on the `isAuthTokenRequired` as detailed in my comment https://github.com/Expensify/App/issues/22133#issuecomment-1620290477

It's plausible that during the tests, downloading the initial image worked as expected, which could be misleading. The issue arises when attempting to download any subsequent image after scrolling through the carousel. This is due to the `isAuthTokenRequired` state not updating accordingly.

Moreover, with the transition of the carousel to its own route, the `onNavigate` function gets invoked immediately to set the initial state for the modal, further highlighting the need for an updated `isAuthTokenRequired` state. I hope this clarification helps!

[the remark this answers is not in the document]
Comment (kidroca) [reply src/components/AttachmentModal.js:107]: Pull Request #20167, which moved the carousel, depended on the proper functioning of the `onNavigate` function. This function is invoked when the carousel mounts and sets the correct value of `isAuthTokenRequired`. While we diligently resolved several merge conflicts when syncing `main` with #20167, it's conceivable that this particular modification related to `isAuthTokenRequired` may have been overlooked. Notably, this PR had received approval a while back, but the merging process was delayed due to a code freeze, which could have further contributed to the oversight.

File: src/components/AttachmentModal.js [hunk 2]
```diff
@@ -78,164 +78,153 @@ const defaultProps = {
...
-            shouldLoadAttachment: false,
-            isAttachmentInvalid: false,
-            isAuthTokenRequired: props.isAuthTokenRequired,
-            attachmentInvalidReasonTitle: null,
-            attachmentInvalidReason: null,
-            source: props.source,
-            modalType: CONST.MODAL.MODAL_TYPE.CENTERED_UNSWIPEABLE,
-            isConfirmButtonDisabled: false,
-            confirmButtonFadeAnimation: new Animated.Value(1),
-            file: props.originalFileName
```
Comment (situchan) [inline src/components/AttachmentModal.js:96]: Coming from https://github.com/Expensify/App/issues/28036:
We could have updated `file` state value whenever `originalFileName` prop changes in useEffect.

Comment (situchan) [reply src/components/AttachmentModal.js:96]: The root cause is [here](https://github.com/Expensify/App/issues/28036#issuecomment-1765250730)

**Pull request merged**